### PR TITLE
feat: notifications

### PR DESCRIPTION
**Pull request by czaja307**
just watering notifications for now, if we need any other kind let me know (prolly conditioning, but we do not have any logic implemented yet)

### CODE REVIEW

File: PlannTer/Backend/Models/PlantModel.swift [hunk 1]
```diff
@@ -198,8 +198,11 @@ class PlantModel: Identifiable, Codable {
         self.prevPrevWateringDate = self.prevWateringDate // zapisz poprzednią datę podlewania
         self.prevWateringDate = Date() // ustaw dzisiejszą datę jako prevWateringDate
         self.nextWateringDate = Calendar.current.date(byAdding: .day, value: wateringFreq ?? 0, to: Date()) // ustaw nextWateringDate na podstawie wateringFreq
+        if let nextWateringDate = nextWateringDate {
+            dispatchNotification(identifier: self.id.description, title: self.name + " is thirsty!", body: "Your plant needs to drink some water.", date: nextWateringDate)
```
Comment (simplyNoOne) [inline PlannTer/Backend/Models/PlantModel.swift:202]: it seems that you only dispatch the notification once the plant already was watered once, but what about just after creating it? add dispatching the notification after the plant was created as well

**Review by simplyNoOne — COMMENTED** [review]
good job overall, notifications look great

**Review by simplyNoOne — CHANGES_REQUESTED** [review]
okey, imma be annoying - you cannot dispatch notifications from a constructor: during the editing and saving we are gonna create a shitload of temporary plants that will never be saved to the database and we do NOT want any notifications for them. This way, said notifications will be created. Rather, add the caling of this method somewhere in the frontend, preferably, as an action happening on pressing the save button. It would be better if we had a way to cancel dispatch, so that when user changes the next watering date or a watering interval, the obsolete notification doesn't get sent, but if it's too much trouble, just leave it as is, but don't create notofications for the temp plants

**Comment by czaja307** [comment]
> okey, imma be annoying - you cannot dispatch notifications from a constructor: during the editing and saving we are gonna create a shitload of temporary plants that will never be saved to the database and we do NOT want any notifications for them. This way, said notifications will be created. Rather, add the caling of this method somewhere in the frontend, preferably, as an action happening on pressing the save button. It would be better if we had a way to cancel dispatch, so that when user changes the next watering date or a watering interval, the obsolete notification doesn't get sent, but if it's too much trouble, just leave it as is, but don't create notofications for the temp plants

dw, reviews are perfect excuse to be annoying
you're absolutely right about this, i didn't thought of temp plants
i'm not sure if handling notifications on frontend is a good practice, but this project is probably swift antipattern by now :p
as for the canceling notifications, i think i already handled that (there can be only one notification for each plant), but maybe i don't see something

gj with reviews, i just hope the next one will get accepted haha

File: PlannTer/Notifications.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    content.body = body
+    content.sound = .default
+    
+    let calendar = Calendar.current
+    let dateComponents = calendar.dateComponents([.year, .month, .day, .hour, .minute], from: date)
+    
+    let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: false)
+    let request = UNNotificationRequest(identifier: identifier, content: content, trigger: trigger)
+    
+    notificationCenter.removePendingNotificationRequests(withIdentifiers: [identifier])
```
Comment (czaja307) [inline PlannTer/Notifications.swift:32]: @simplyNoOne this should do the work, i suppose

**Review by simplyNoOne — APPROVED** [review]
good job

**Pull request merged**